src/Status.cxx

```diff
@@ -2,120 +2,127 @@
+  if (TimeDifference.count() < 1e-10) {
     return std::pair<double, double>({0, 0});
   }
-  std::pair<double, double> result{
-      average(value.Mbytes().first, time_difference),
-      standard_deviation(value.Mbytes().first, value.Mbytes().second,
-                         time_difference)};
+  std::pair<double, double> result(
+      1e3 * average(Value.getMbytes().first, TimeDifference.count()),
+      1e3 * standardDeviation(Value.getMbytes().first, Value.getMbytes().second,
+                              TimeDifference.count()));
   return result;
 }
 
-// data-race safe += with double
-template <class T> void atomic_add(std::atomic<T> &value, const T &other) {
-  T current = value.load();
-  while (!value.compare_exchange_weak(current, current + other)) {
-    current = value.load();
+/// Implement atomic_add for a generic type. The type must implement operator +
+/// and =
+template <class T> void atomic_add(std::atomic<T> &Value, const T &Other) {
+  T Current = Value.load();
+  while (!Value.compare_exchange_weak(Current, Current + Other)) {
+    Current = Value.load();
   }
 }
 
 const FileWriter::Status::MessageInfo &FileWriter::Status::MessageInfo::
-operator=(const FileWriter::Status::MessageInfo &other) {
-  Mbytes_.store(other.Mbytes_.load());
-  Mbytes_square_.store(other.Mbytes_square_.load());
-  messages_.store(other.messages_.load());
-  messages_square_.store(other.messages_square_.load());
-  errors_.store(other.errors_.load());
+operator=(const FileWriter::Status::MessageInfo &Other) {
+  Mbytes.store(Other.Mbytes.load());
+  MbytesSquare.store(Other.MbytesSquare.load());
+  Messages.store(Other.Messages.load());
+  MessagesSquare.store(Other.MessagesSquare.load());
+  Errors.store(Other.Errors.load());
   return *this;
 }
 
 const FileWriter::Status::MessageInfo &FileWriter::Status::MessageInfo::
-operator+=(const FileWriter::Status::MessageInfo &other) {
-  atomic_add(Mbytes_, other.Mbytes_.load());
-  atomic_add(Mbytes_square_, other.Mbytes_square_.load());
-  atomic_add(messages_, other.messages_.load());
-  atomic_add(messages_square_, other.messages_square_.load());
-  atomic_add(errors_, other.errors_.load());
+operator+=(const FileWriter::Status::MessageInfo &Other) {
+  atomic_add(Mbytes, Other.Mbytes.load());
+  atomic_add(MbytesSquare, Other.MbytesSquare.load());
+  atomic_add(Messages, Other.Messages.load());
+  atomic_add(MessagesSquare, Other.MessagesSquare.load());
+  atomic_add(Errors, Other.Errors.load());
   return *this;
 }
 
 const FileWriter::Status::MessageInfo &
-FileWriter::Status::MessageInfo::message(const double &message_size) {
-  std::lock_guard<std::mutex> lock(mutex_);
-  auto size = message_size * 1e-6;
-  atomic_add(Mbytes_, size);
-  atomic_add(Mbytes_square_, size * size);
-  atomic_add(messages_, 1.0);
-  atomic_add(messages_square_, 1.0);
+FileWriter::Status::MessageInfo::message(const double &MessageSize) {
+  std::lock_guard<std::mutex> Lock(Mutex);
```

Comment:
In principle we can live without the lock, nevertheless in that case there could be inconsistencies. Since ``message`` (and ``error``) method and ``+=`` operator are used in different threads the data can be out of sync. Of course we can live with that, could happen that sometimes numbers will be a little silly.
I discussed this (and the next) comment with Dominik: he points out that in order to be completely consistent we need to use the same lock in ``operator+=`` and ``Report::produce_single_report``. 
Are you in favour of using the locks and trade a (possibly negligible, to be evaluated) slowdown with consistency in the statistic or no lock and possibility of inconsistencies? Of course these  inconsistencies are not critical at all, the report we produce is a just plus.

Reply:
The only operations the lock protects here are `atomic_add`, so are you saying `atomic_add` is not actually atomic?

Reply:
The ``atomic_add`` has no inconsistencies, what I'm worried about is the lack of synchronization among the different variables which are used to compute the messages statistic. I'll try to explain myself with an example:
 
1. thread1 consumes a message and calls  ``MessageInfo::message``. At the same time thread 2 is sunning ``Report::produce_single_report``;
2. T1 executes ``atomic_add`` on ``Mbytes`` and ``MbytesSquare``. T2 executes ``StreamMasterInfo::add`` on the same data;
3. before T1 executes ``atomic_add`` on ``Messages`` T2 is executes  ``atomic_add`` on all the fields.
4. while T1 finishes all the  ``atomic_add`` T2 produces a message with some values at the current "iteration" and some at the previous.

Now, there is no inconsistency on the ``atomic_add`` itself but among the "iteration" at which the different values are considered.
Even worse: in the case above (without the locks) T2 could call ``MessageInfo::reset`` before T1 executes ``atomic_add`` on  ``Messages`` and ``MessagesSquare``.

Reply:
That's true, but all these operations are additions so it won't lead to a large cumulative error, right? If so, maybe it is better to prioritise performance of the statistics code over it being completely accurate? I don't know the requirements for the status information so I leave the decision to you.

Reply:
When we looked at the code, I suggested adding the required locks also on the reader side in order to calculate consistent values, because it's easier to remove locks later if it ever turns out to be a bottleneck, than to add locks in retrospect.
If our goal is only some form of eventual consistency, then like you said we would not need any locks.
The update frequency of the statistics is quite low, so my feeling is that we can afford the price for the consistency. If benchmarks tell otherwise, we can still lower our consistency standards.
So I would vote for keeping the locks at the moment.